app/src/main/java/com/azuresamples/msalnativeauthandroidkotlinsampleapp/MFAVerificationFragment.kt

```diff
@@ -0,0 +1,151 @@
+        binding.verifyChallenge.setOnClickListener {
+            verifyChallenge()
+        }
+
+        binding.resendChallengeText.setOnClickListener {
+            resendChallenge()
+        }
+    }
+
+    private fun verifyChallenge() {
+        CoroutineScope(Dispatchers.Main).launch {
+            val emailCode = binding.challengeText.text.toString()
+
+            val actionResult = currentState.submitChallenge(emailCode)
+
+            when (actionResult) {
+                is SignInResult.Complete -> {
+                    Toast.makeText(
+                        requireContext(),
+                        getString(R.string.sign_in_successful_message),
+                        Toast.LENGTH_SHORT
+                    ).show()
+                    finish()
+                }
+                is SubmitChallengeError -> {
+                    handleSubmitChallengeError(actionResult)
+                }
+                else -> {
+                    displayDialog(getString(R.string.unexpected_sdk_result_title), actionResult.toString())
+                }
+            }
+        }
+    }
+
+    private fun resendChallenge() {
+        clearChallengeText()
+
+        CoroutineScope(Dispatchers.Main).launch {
+            val actionResult = currentState.requestChallenge()
+
+            when (actionResult) {
+                is MFARequiredResult.VerificationRequired -> {
+                    currentState = actionResult.nextState
+                    Toast.makeText(requireContext(), getString(R.string.resend_challenge_message), Toast.LENGTH_LONG).show()
+                }
+                is MFAError -> {
+                    handleMFAError(actionResult)
+                }
+                else -> {
+                    displayDialog(getString(R.string.unexpected_sdk_result_title), actionResult.toString())
+                }
+            }
+        }
+    }
+
+    private fun clearChallengeText() {
+        binding.challengeText.text?.clear()
+    }
+
+    fun displayDialog(error: String? = null, message: String?) {
+        val builder = AlertDialog.Builder(context)
+        builder.setTitle(error)
+            .setMessage(message)
+        val alertDialog = builder.create()
+        alertDialog.show()
+    }
+
+    private fun handleSubmitChallengeError(error: SubmitChallengeError) {
+        when {
+            error.isInvalidChallenge()
+            -> {
+                displayDialog(error.error, error.errorMessage)
+            }
+            else -> {
+                displayDialog(getString(R.string.unexpected_sdk_result_title), error.errorMessage)
+            }
+        }
+    }
+
+    private fun handleMFAError(error: MFAError) {
+        when {
+            error.isError()
+            -> {
+                displayDialog(error.error, error.errorMessage)
+            }
+            else -> {
+                displayDialog(getString(R.string.unexpected_sdk_result_title), error.errorMessage)
```

Comment:
`error.exception.message`? Or does that not apply here?

Reply:
Done